x-pack/metricbeat/module/meraki/http.go

```diff
@@ -0,0 +1,69 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package meraki
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/tomnomnom/linkheader"
+)
+
+type paginator[T any] struct {
+	setStart  func(string)
+	doRequest func() (T, *resty.Response, error)
+	onError   func(error, *resty.Response) error
+	onSuccess func(T) error
+}
+
+func NewPaginator[T any](
+	setStart func(string),
+	doRequest func() (T, *resty.Response, error),
+	onError func(error, *resty.Response) error,
+	onSuccess func(T) error,
+) *paginator[T] {
+	return &paginator[T]{
+		setStart:  setStart,
+		doRequest: doRequest,
+		onError:   onError,
+		onSuccess: onSuccess,
+	}
+}
+
+func (p *paginator[T]) GetAllPages() error {
+	hasMorePages := true
+
+	for hasMorePages {
+		val, res, err := p.doRequest()
```

Comment:
Would passing ctx here be helpful ?
To cancel long-running pagination

Reply:
this would be good, but i don't think we have any way to actually use it in the metricset. there's no cancelable context passed to `Fetch` that we could pass down to this function.

Reply:
this would be good, but i don't think we have any way to actually use it in the metricset. there's no cancelable context passed to `Fetch` that we could pass down to this function.